pkg/kubelet/pleg/generic.go

```diff
@@ -218,36 +226,59 @@ func (g *GenericPLEG) relist() {
 		needsReinspection = make(map[types.UID]*kubecontainer.Pod)
 	}
 
+	// update PodCache and PodRecords
+	g.updatePodCacheByWorkers(eventsByPodID, needsReinspection)
+
+	if g.cacheEnabled() {
+		// reinspect any pods that failed inspection during the previous relist
+		if len(g.podsToReinspect) > 0 {
+			glog.V(5).Infof("GenericPLEG: Reinspecting pods that previously failed inspection")
+			for pid, pod := range g.podsToReinspect {
+				if err := g.updateCache(pod, pid); err != nil {
+					glog.Errorf("PLEG: pod %s/%s failed reinspection: %v", pod.Name, pod.Namespace, err)
+					needsReinspection[pid] = pod
+				}
+			}
+		}
+
+		// Update the cache timestamp.  This needs to happen *after*
+		// all pods have been properly updated in the cache.
+		g.cache.UpdateTime(timestamp)
+	}
+
+	// make sure we retain the list of pods that need reinspecting the next time relist is called
+	g.podsToReinspect = needsReinspection
+}
+
+func (g *GenericPLEG) updatePodCacheByWorkers(eventsByPodID map[types.UID][]*PodLifecycleEvent,
```

Comment:
It's cleaner if the function does only one thing. You can return a list of "updated" pods, and use that to deal with the reinspection logic. That way, you don't need to pass the `needsReinspection` map at all.